Framework/API/src/WorkspaceGroup.cpp

```diff
@@ -278,21 +279,32 @@ void WorkspaceGroup::workspaceDeleteHandle(
 }
 
 /**
- * Callback when a after-replace notification is received
- * Replaces a member if it was replaced in the ADS.
- * @param notice :: A pointer to a workspace after-replace notificiation object
+ * Callback when a before-replace notification is received
+ * Replaces a member if it was replaced in the ADS and checks
+ * for duplicate members within the group
+ * @param notice :: A pointer to a workspace before-replace notification object
  */
-void WorkspaceGroup::workspaceReplaceHandle(
+void WorkspaceGroup::workspaceBeforeReplaceHandle(
     Mantid::API::WorkspaceBeforeReplaceNotification_ptr notice) {
   std::lock_guard<std::recursive_mutex> _lock(m_mutex);
 
-  const std::string replacedName = notice->objectName();
+  const auto oldObject = notice->oldObject();
+
   for (auto &workspace : m_workspaces) {
-    if ((*workspace).name() == replacedName) {
+    if (workspace == oldObject) {
       workspace = notice->newObject();
       break;
     }
   }
+
+  // Remove any duplicate pointers
+  std::sort(m_workspaces.begin(), m_workspaces.end());
+  auto endIter =
+      std::unique(m_workspaces.begin(), m_workspaces.end(),
```

Comment:
Good point - I was originally getting the raw pointers and forgot to remove that lambda when I realised it wasn't needed.
I am just figure out what is causing the unit tests below to fail and then this should be pretty much there unless there is any other cases either of us can think of?